src/Adapter/Product/Repository/CombinationRepository.php

```diff
@@ -0,0 +1,253 @@
+
+        return new CombinationId($id);
+    }
+
+    /**
+     * @param int $productId
+     * @param int $limit
+     * @param int $offset
+     * @param array $filters
+     *
+     * @return array<int, array<string, mixed>>
+     */
+    public function getProductCombinations(int $productId, ?int $limit = null, ?int $offset = null, array $filters = []): array
+    {
+        $qb = $this->getCombinationsQueryBuilder($productId, $filters)
+            ->select('pa.*')
+            ->setParameter('productId', $productId)
+        ;
+
+        if ($offset) {
+            $qb->setFirstResult($offset);
+        }
+
+        if ($limit) {
+            $qb->setMaxResults($limit);
+        }
+
+        return  $qb->execute()->fetchAll();
+    }
+
+    /**
+     * @param int $productId
+     * @param array $filters
+     *
+     * @return int
+     */
+    public function getTotalCombinationsCount(int $productId, array $filters = []): int
+    {
+        $qb = $this->getCombinationsQueryBuilder($productId, $filters)->select('COUNT(pa.id_product_attribute) AS total_combinations');
+
+        return (int) $qb->execute()->fetch()['total_combinations'];
+    }
+
+    /**
+     * @param int[] $combinationIds
+     * @param LanguageId $langId
+     *
+     * @return array<int, array<int, mixed>>
+     */
+    public function getAttributesInfoByCombinationIds(array $combinationIds, LanguageId $langId): array
+    {
+        $attributeCombinationAssociations = $this->getAttributeCombinationAssociations($combinationIds);
+
+        $attributeIds = array_unique(array_map(function ($attributeByCombination) {
+            return $attributeByCombination['id_attribute'];
+        }, $attributeCombinationAssociations));
+
+        $attributesInfoByAttributeId = $this->getAttributesInformation($attributeIds, $langId->getValue());
+
+        $attributesInfoByCombinationId = [];
+        foreach ($attributeCombinationAssociations as $attributeCombinationAssociation) {
+            $combinationId = (int) $attributeCombinationAssociation['id_product_attribute'];
+            $attributeId = (int) $attributeCombinationAssociation['id_attribute'];
+            $attributesInfoByCombinationId[$combinationId] = $attributesInfoByAttributeId[$attributeId];
+        }
+
+        return $attributesInfoByCombinationId;
+    }
+
+    /**
+     * @param CombinationId $combinationId
+     * @param int[] $attributeIds
+     */
+    public function saveProductAttributeAssociation(CombinationId $combinationId, array $attributeIds): void
+    {
+        $attributesList = [];
+        foreach ($attributeIds as $attributeId) {
+            $attributesList[] = [
+                'id_product_attribute' => $combinationId->getValue(),
+                'id_attribute' => $attributeId,
+            ];
+        }
+
+        try {
+            if (!Db::getInstance()->insert('product_attribute_combination', $attributesList)) {
```

Comment:
Maybe you should check if it exists before?

Reply:
You unresolved this one so I guess it doesn't seem relevant to you? ^^

I mean this to avoid doublons or insertion errors in case the association already exist, then the present associations should be ignored to avoid making all the other insertions fail

Reply:
Oh my bad, actually you made a modif to check that all attributes exist I didn't think of that one but it's even safer indeed I still worry about doublons in relationship rows though That's what my initial comment was about but i understand now it was not clear enough ^^